src/shims/mod.rs

```diff
@@ -27,9 +28,74 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
         }
         // There are some more lang items we want to hook that CTFE does not hook (yet).
         if this.tcx.lang_items().align_offset_fn() == Some(instance.def.def_id()) {
-            // FIXME: return a real value in case the target allocation has an
-            // alignment bigger than the one requested.
-            let n = u128::max_value();
+            let ptr = this.read_scalar(args[0])?.not_undef()?;
+            let align = this.read_scalar(args[1])?.not_undef()?;
+
+            let n = {
+                let p = this.force_bits(ptr, this.pointer_size())?;
+                let a = this.force_bits(align, this.pointer_size())?;
+
+                let stride = this.memory().get(ptr.assert_ptr().alloc_id)?.align.bytes() as u128;
+                // if the allocation alignment is at least the required alignment, we use the
```

Comment:
instead of rerolling the impl, could we just call the libcore impl? Essentially `((stride + ptr.offset) as *const ()).align_offset(align) - stride` should give you the number to return if `stride >= a`

Reply:
Yeah, we definitely should not copy-paste that code.

Reply:
This works with a `saturating_sub` for the case when the align offset is 0 for the `*const ()`

Reply:
:man_facepalming: 	oops, sorry. You don't need the subtraction at all. I was assuming that the `align_offset` returned the aligned pointer, but it returns the offset to add to the original pointer, so the subtraction is actually wrong (which is probably why you were getting panics)